crates/miden-objects/src/account/component/metadata/storage/type_registry.rs

```diff
@@ -0,0 +1,615 @@
+    },
+    #[error("word type ` {0}` not found in the type registry")]
+    WordTypeNotFound(SchemaTypeIdentifier),
+}
+
+impl SchemaTypeError {
+    /// Creates a [`SchemaTypeError::ParseError`].
+    pub fn parse(
+        input: impl Into<String>,
+        schema_type: SchemaTypeIdentifier,
+        source: impl Error + Send + Sync + 'static,
+    ) -> Self {
+        SchemaTypeError::ParseError {
+            input: input.into(),
+            schema_type,
+            source: Box::new(source),
+        }
+    }
+}
+
+// SCHEMA TYPE
+// ================================================================================================
+
+/// A newtype wrapper around a `String`, representing a schema type identifier.
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+#[cfg_attr(feature = "std", derive(::serde::Deserialize, ::serde::Serialize))]
+#[cfg_attr(feature = "std", serde(transparent))]
+pub struct SchemaTypeIdentifier(String);
+
+impl SchemaTypeIdentifier {
+    /// Creates a new [`SchemaTypeIdentifier`] from a `String`.
+    ///
+    /// The name must follow a Rust-style namespace format, consisting of one or more segments
+    /// (non-empty, and alphanumerical) separated by double-colon (`::`) delimiters.
+    ///
+    /// # Errors
+    ///
+    /// - If the identifier is empty.
+    /// - If any segment is empty or contains something other than alphanumerical
+    ///   characters/underscores.
+    pub fn new(s: impl Into<String>) -> Result<Self, SchemaTypeError> {
+        let s = s.into();
+        if s.is_empty() {
+            return Err(SchemaTypeError::InvalidTypeName(
+                s.clone(),
+                "schema type identifier is empty".to_string(),
+            ));
+        }
+        for segment in s.split("::") {
+            if segment.is_empty() {
+                return Err(SchemaTypeError::InvalidTypeName(
+                    s.clone(),
+                    "empty segment in schema type identifier".to_string(),
+                ));
+            }
+            if !segment.chars().all(|c| c.is_ascii_alphanumeric() || c == '_') {
+                return Err(SchemaTypeError::InvalidTypeName(
+                    s.clone(),
+                    format!("segment '{segment}' contains invalid characters"),
+                ));
+            }
+        }
+        Ok(Self(s))
+    }
+
+    /// Returns the schema type identifier for the native [`Felt`] type.
+    pub fn native_felt() -> SchemaTypeIdentifier {
+        SchemaTypeIdentifier::new("felt").expect("type is well formed")
+    }
+
+    /// Returns the schema type identifier for the native [`Word`] type.
+    pub fn native_word() -> SchemaTypeIdentifier {
+        SchemaTypeIdentifier::new("word").expect("type is well formed")
+    }
+
+    /// Returns the schema type identifier for storage map slots.
+    pub fn storage_map() -> SchemaTypeIdentifier {
+        SchemaTypeIdentifier::new("map").expect("type is well formed")
+    }
```

Comment:
Nit: Should this actually be `storage_map` to disambiguate with other potential map types?